nav2_simple_commander/nav2_simple_commander/line_iterator.py

```diff
@@ -0,0 +1,201 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This is a Python3 API for a line iterator.
+
+It provides the ability to iterate through the points of a line.
+"""
+
+
+from cmath import sqrt
+
+
+class LineIterator():
+    def __init__(self, x0, y0, x1, y1, step_size=1.0):
+        """Initializer for LineIterator.
+
+        Args:
+            x0 (Float): Abscissa of the initial point
+            y0 (Float): Ordinate of the initial point
+            x1 (Float): Abscissa of the final point
+            y1 (Float): Ordinate of the final point
+            step_size (Float, optional): Resolution of increments. Defaults to 1.
+        """
+        self.x0_ = x0
+        self.y0_ = x0
+        self.x1_ = x1
+        self.y1_ = y1
+        self.x_ = x0
+        self.y_ = y0
+        self.step_size_ = step_size
+
+        if x1 != x0 and y1 != y0:
+            self.valid_ = True
+            self.m_ = (y1-y0)/(x1-x0)
+            self.b_ = y1 - (self.m_*x1)
+            if (self.b_ < 0):
+                self.equation_ = "y = " + str(self.m_) + "*x " + str(self.b_)
+            else:
+                self.equation_ = "y = " + str(self.m_) + "*x + " + str(self.b_)
+        elif x1 == x0 and y1 != y0:
+            self.valid_ = True
+            self.equation_ = "x = " + str(x1)
+        elif y1 == y1 and x1 != x0:
+            self.valid_ = True
+            self.m_ = (y1-y0)/(x1-x0)
+            self.b_ = y1 - (self.m_*x1)
+            self.equation_ = "y = " + str(y1)
+        else:
+            self.valid_ = False
+            self.equation_ = "Invalid"
+
+    def isValid(self):
+        """Returns True if the line is valid
+        Returns:
+            Bool: Flag if live is valid (true) or not (false)
+        """
+        return self.valid_
+
+    def advance(self):
+        """Advances to the next point in the line."""
+        if self.x1_ > self.x0_:
+            if self.x_ < self.x1_:
+                self.x_ = round(self.clamp(
+                    self.x_ + self.step_size_, self.x0_, self.x1_), 5)
+                self.y_ = round(self.m_ * self.x_ + self.b_, 5)
+            else:
+                self.valid_ = False
+        elif self.x1_ < self.x0_:
+            if self.x_ > self.x1_:
+                self.x_ = round(self.clamp(
+                    self.x_ - self.step_size_, self.x1_, self.x0_), 5)
+                self.y_ = round(self.m_ * self.x_ + self.b_, 5)
+            else:
+                self.valid_ = False
+        else:
+            if self.y1_ > self.y0_:
+                if self.y_ < self.y1_:
+                    self.y_ = round(self.clamp(
```

Comment:
And X?

Reply:
This if statement handles the case where the line is vertical (X1 == X0), so the equation will be for example x = 3 and only Y is changing. Thus I'm only incrementing Y here as X is fixed in the vertical line

Reply:
Then why compute / store `m` and `b`?

Reply:
In the case of X1 == X0 (Line 54), m and b are not computed nor stored, and they are not used in the advance function in this case (Lines 90 --> 93)

Reply:
m and b are only computed/stored only in the cases where the line equation contains m and b (all cases other than the vertical line case)